artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -586,10 +586,24 @@ pub fn byteindex(
     offset: Option<Value>,
 ) -> Result<Value, Error> {
     #[cfg(feature = "core-regexp")]
-    if let Ok(_pattern) = unsafe { Regexp::unbox_from_value(&mut substring, interp) } {
-        return Err(NotImplementedError::from("String#byteindex with Regexp pattern").into());
-    }
     let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    if let Ok(pattern) = unsafe { Regexp::unbox_from_value(&mut substring, interp) } {
+        let offset_number = match implicitly_convert_to_int(interp, offset.into()) {
+            Err(_) => None,
+            Ok(value) => Some(value),
+        };
+
+        let mut matches = pattern.match_(interp, Some(s.as_slice()), offset_number, None)?;
+        if matches.is_nil() {
+            return Ok(Value::nil());
+        }
+        let first_ocurrence = unsafe { MatchData::unbox_from_value(&mut matches, interp)? };
+        let ocurrence_index = first_ocurrence.begin(matchdata::Capture::GroupIndex(0)).unwrap();
```

Comment:
I was a bit confused here between line 596 and 601 on what `offset_number` meant. From playing around with this API (and what you've confirmed with the integration tests), `offset_number` is a byte offset into the "haystack" (string to search). It is NOT the capturing group to use.

This example made this clear to me:

```console
[3.2.2] > s = "abcdef"
=> "abcdef"
[3.2.2] > s.byteindex(/(c).*(f)/, 2)
=> 2
```

Could you add some comments to this code and the above case to `string_test.rb` to document this bit of the API?